### ISSUE TITLE
Ansible 2.5.1:  'ansible_virtualization_type' is undefined

### ISSUE THREAD

**Issue by Solbadguy** [StreisandEffect/streisand#1285]
**FAILED! => {"msg": "The conditional check 'ansible_virtualization_type != 'lxc'' failed. The error was: error while evaluating conditional (ansible_virtualization_type != 'lxc'): 'ansible_virtualization_type' is undefined\n\nThe error appears to have been in '/home/sol/streisand/playbooks/roles/sysctl/tasks/main.yml': line 3, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n- block:\n    - name: Apply custom sysctl values\n      ^ here\n"}**

Installed on the minimum tariff plan of Hetzner.
During the installation, I tried variants 7 and 8 - both ended in error with a mention of the virtualization error LXC.

hetzner VPS 2gb, 20gb hdd
ubuntu 16.04
ansible 2.5.1
python version = 2.7.12 
git rev-parse HEAD
f9575348493438431b27ee2e83a84033eb2b3b6e

**Comment by ruvhell** [comment]
google cloud
ubuntu 16.04
on TASK [sysctl : Apply custom sysctl values]
same error

**Comment by ruvhell** [comment]
Manually add sysctl key to conf.
iam del 
- name: "Enable IPv4 traffic forwarding"
  sysctl:
    name: net.ipv4.ip_forward
    value: 1
  when: ansible_virtualization_type != 'lxc'
from streisand/playbooks/roles/ip-forwarding/tasks/main.yml

- name: Apply the custom sysctl values
  include_role:
    name: sysctl
from streisand/playbooks/roles/common/tasks/main.yml

and error go away

**Comment by Egrace** [comment]
Same problem. Does somebody know how to fix it?

**Comment by Escinelio** [comment]
Confirm.
The same error.
Hetzner
ansible 2.5.1
python version = 2.7.12

**Comment by cpu** [comment]
Hi folks,

I won't have a chance to look at this until ~tomorrow at the earliest. Based on the reported Ansible versions in this thread my first guess is that this is a regression caused by Ansible 2.5.1.

@Escinelio @Egrace @ruvhell @Solbadguy Can one of you please try uninstalling Ansible and then installing `2.5.0` instead of `2.5.1` and seeing if you encounter the same problem?

**Comment by dvkudr** [comment]
Hi gyus!
I experienced the same issue as one described by the topic starter.
I've tried the suggested workaround (i.e. installing Ansible 2.5.0 instead of 2.5.1) and it seems that it's sorted the problem out.

**Comment by cpu** [comment]
@dvkudr Thanks for confirming that downgrading to 2.5.0 works for the short term. We'll figure out a solution to the root issue ASAP.

**Comment by nopdotcom** [comment]
Poking at this a little more:

2.5.1: Works for EC2 install from OS X; works on a localhost Vultr. Fails on GCE.
2.5.0: Works on GCE.

I put the broken 2.5.1 and working 2.5.0 hosts in an inventory. There's no obvious difference in virtualization facts:

```
$ ansible -m setup -a 'filter=ansible_virtualization_*' gce
192.0.10.251 | SUCCESS => {
    "ansible_facts": {
        "ansible_virtualization_role": "NA",
        "ansible_virtualization_type": "NA"
    },
    "changed": false
}
$ ansible -m setup -a 'filter=ansible_virtualization_*' gce_250
192.0.10.250 | SUCCESS => {
    "ansible_facts": {
        "ansible_virtualization_role": "NA",
        "ansible_virtualization_type": "NA"
    },
    "changed": false
}
```

**Comment by nopdotcom** [comment]
Browsing through Ansible's issue lists I see https://github.com/ansible/ansible/issues/39049 and https://github.com/ansible/ansible/pull/38302 , which *sound* like they should be affecting us.

My guess is:

* Document "Use Ansible 2.5.0" in `README.md`;

* Add `ansible[azure]==2.5.0` to `requirements.txt`

* ...revert both when 2.5.2 ships?

**Comment by Escinelio** [comment]
@dvkudr 
2.5.1/2.5.0/2.4.0 fails on Scaleway, Hetzner
2.4.1 works fine for both

**Comment by nopdotcom** [comment]
I'm confused, and I think we need to split up this bug. 

@dvkudr says Ansible 2.5.1 broken, 2.5.0 working. (no provider mentioned)

@ruvhell says Ansible 2.5.1 broken on GCE.

I say Ansible 2.5.1 broken on GCE, 2.5.0 working.

@Solbadguy says Hetzner broken at 2.5.1.

@Escinelio says Scaleway/Hetzner broken on both 2.5.1, 2.5.0, but working on 2.4.1.

We didn't really go through this with a proper bug template and I'm losing track of what's broken for whom. I'm going to declare this issue to be limited to those providers working at 2.5.0 and broken at 2.5.1. If **reverting to Ansible 2.5.0** doesn't work for you, please open a separate issue so we can track down details.

@Escinelio, could you open two issues for the problems you're seeing on Scaleway and Hetzner?

Scaleway has had some problems in the past, so I'd like to keep it from getting mixed up with the others.

**Comment by cpu** [comment]
Thanks for trying to sort through this @nopdotcom.

> Document "Use Ansible 2.5.0" in README.md;
> Add ansible[azure]==2.5.0 to requirements.txt
> ...revert both when 2.5.2 ships?

This sounds good to me :+1: - @nopdotcom Do you have time to make the PR for :point_up: ?

>  If reverting to Ansible 2.5.0 doesn't work for you, please open a separate issue so we can track down details.

Seconded. I'm going to update the title of this issue to reflect this.

> @Escinelio, could you open two issues for the problems you're seeing on Scaleway and Hetzner?

Scaleway is unsupported & has lots of issues. Please do not open an issue for Scaleway problems.

I strongly recommend folks interested in a stable, novice friendly Streisand use a **supported provider** (DigitalOcean, Linode, AWS, GCE, etc). If you are using one of the **advanced** provisioning modes (localhost, existing server) on an unsupported provider the expectation is that you have to take the lead in your own debugging, PRs welcome!

**Comment by Solbadguy** [comment]
ansible 2.4.1 - Working for Hetzner

**Comment by saethydd** [comment]
python 2.7.12, tried ansible 2.4.1, 2.5.0. and 2.5.1. neither of them worked with sysctl properly
kernel 4.4.0 (?), openvz/lxc, provided by vpsnet.lt

**Comment by nopdotcom** [comment]
Thanks for the continued feedback on versions. I think we've got this one nailed down now, although there are some other cans of worms opened.

People using the `./util/venv-dependencies.sh` script from Stresisand are now getting Ansible 2.5.0 instead of 2.5.1.  Ansible 2.5.2 shipped to pypi earlier today, so people running an unversioned `pip install ansible` will get a working version. I'm going to remove my Ansible version pin from `requirements.txt` in a couple days, once things calm down. 

This leaves Homebrew Ansible at 2.5.1 until they update, but there are other problems with Ansible installed through Homebrew...like, [they ship all their own modules anyway](https://github.com/Homebrew/homebrew-core/blob/master/Formula/ansible.rb#L22). I'll file an issue against us using them. 

@saethydd, I'm not sure if that provider is going to work. I think we'd need to look at that on a fresh issue, if getting it working is possible at all.

**Comment by nopdotcom** [comment]
Yeah, it's https://github.com/ansible/ansible/issues/38947 and I'm not sure what to do about it. Any ideas?

**Comment by gleb108** [comment]
I've got the same issue with Hetzner server with ansible 2.5.2
As temporarily quick and dirty solution I've disable this checks and it worked fine for me.
Probably it'll be help for someone:

diff --git a/playbooks/roles/ip-forwarding/tasks/main.yml b/playbooks/roles/ip-forwarding/tasks/main.yml
index f491ef5..996bf46 100644
--- a/playbooks/roles/ip-forwarding/tasks/main.yml
+++ b/playbooks/roles/ip-forwarding/tasks/main.yml
@@ -3,7 +3,7 @@
   sysctl:
     name: net.ipv4.ip_forward
     value: 1
-  when: ansible_virtualization_type != 'lxc'
+#  when: ansible_virtualization_type != 'lxc'
 
 - name: "Add IPv4 traffic forwarding persistence service to init"
   copy:
diff --git a/playbooks/roles/shadowsocks/tasks/main.yml b/playbooks/roles/shadowsocks/tasks/main.yml
index 38064b9..8e20786 100644
--- a/playbooks/roles/shadowsocks/tasks/main.yml
+++ b/playbooks/roles/shadowsocks/tasks/main.yml
@@ -4,7 +4,7 @@
     name: net.ipv4.tcp_fastopen
     value: 3
     state: present
-  when: ansible_virtualization_type != 'lxc'
+#  when: ansible_virtualization_type != 'lxc'
 
 - name: Add the Shadowsocks PPA
   apt_repository:
diff --git a/playbooks/roles/sysctl/tasks/main.yml b/playbooks/roles/sysctl/tasks/main.yml
index 17b77ee..ea8f9a9 100644
--- a/playbooks/roles/sysctl/tasks/main.yml
+++ b/playbooks/roles/sysctl/tasks/main.yml
@@ -9,4 +9,4 @@
         reload: yes
         state: present
       with_items: "{{ sysctl_values }}"
-  when: ansible_virtualization_type != 'lxc'
+#  when: ansible_virtualization_type != 'lxc'

**Comment by stufently** [comment]
Any solution for last stable ansible ?

**Comment by nopdotcom** [comment]
I think the upgrade to 2.5.4 fixes it, so I'm claiming closed via #1368.
